tests/vmi_vsock_test.go

```diff
@@ -147,4 +151,75 @@ var _ = Describe("[Serial][sig-compute]VSOCK", func() {
 			Expect(domSpec2.Devices.VSOCK.CID.Address).To(Equal(*vmi2.Status.VSOCKCID))
 		})
 	})
+
+	Context("API access", func() {
+		It("should communicate with VMI via VSOCK", func() {
+			virtClient, err := kubecli.GetKubevirtClient()
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Creating a VMI with VSOCK enabled")
+			vmi := tests.NewRandomFedoraVMI()
+			vmi.Spec.Domain.Devices.AutoattachVSOCK = pointer.Bool(true)
+			vmi = tests.RunVMIAndExpectLaunch(vmi, 60)
+
+			By("Logging in as root")
+			err = console.LoginToFedora(vmi)
+			Expect(err).ToNot(HaveOccurred())
+
+			By("Starting a server on guest via VSOCK")
+			port := 8888
+			tests.StartPythonVsockServer(vmi, *vmi.Status.VSOCKCID, port)
+
+			By("Connect to the guest via API")
+			cliConn, svrConn := net.Pipe()
+			stopChan := make(chan error)
+			respChan := make(chan string)
+			go func() {
+				defer GinkgoRecover()
+				vsock, err := virtClient.VirtualMachineInstance(vmi.Namespace).VSOCK(vmi.Name, &v1.VSOCKOptions{TargetPort: uint32(port)})
+				if err != nil {
+					stopChan <- err
+				}
+				stopChan <- vsock.Stream(kubecli.StreamOptions{
+					In:  svrConn,
+					Out: svrConn,
+				})
+			}()
+			By("Writing to the Guest")
+			message := "Hello World!"
+			go func() {
+				defer GinkgoRecover()
+				message := "Hello World!"
+				_, err = cliConn.Write([]byte(message))
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			By("Reading from the Guest")
+			go func() {
+				defer GinkgoRecover()
+				buf := make([]byte, 1024, 1024)
+				// reading qemu vnc server
+				n, err := cliConn.Read(buf)
+				Expect(err).NotTo(HaveOccurred())
+				respChan <- string(buf[0:n])
+			}()
+
+			select {
+			case resp := <-respChan:
+				Expect(resp).To(Equal(message))
+			case <-time.After(1 * time.Minute):
+				Fail("Timout communicate with the Vsock server in Guest.")
+			}
+
+			By("Close the stream")
+			err = cliConn.Close()
```

Comment:
There may be cases where we are not properly closing the connection. Since it is just one test and the VM get's deleted anyway (where the connection then definitely breaks), it may not matter too much.

Reply:
I may not get this. Do you mean that even the `cliConn.Close()` doesn't return error, the connection still can be closed improperly? If that's the case, what's the suggestion?

Reply:
No, I just misread the code. It would be fine. But could you make it a defer after you got the connection back? There are right now just go routines started, but if that changes we may return without a close.

Reply:
Got it. Here I want to test the client correctly close the connection case. In terms of any failure occurs before, I added a defer block to close the connect at last.